### PR TITLE
[validation] Two small clang lock annotation improvements

### PR DESCRIPTION
**Pull request by amitiuttarwar**
Based on reviewing #21188 

the first commit switches the lock annotations on `CheckInputScripts` to be on the function declaration instead of on the function definition. this ensures that all call sites are checked, not just ones that come after the definition. 

the second commit adds a note to the developer-notes section to clarify where the annotations should be applied.

### CODE REVIEW

**Comment by amitiuttarwar** [comment]
to test this out, reviewers can apply this diff locally

```diff
- bool CheckInputScripts(const CTransaction& tx, TxValidationState& state, const CCoinsViewCache& inputs, unsigned int flags, bool cacheSigStore, bool cacheFullScriptStore, PrecomputedTransactionData& txdata, std::vector<CScriptCheck>* pvChecks)
+ RecursiveMutex cs_new;
+ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state, const CCoinsViewCache& inputs, unsigned int flags, bool cacheSigStore, bool cacheFullScriptStore, PrecomputedTransactionData& txdata, std::vector<CScriptCheck>* pvChecks) EXCLUSIVE_LOCKS_REQUIRED(cs_new)
```

and see that the compiler warns about the call site from `ConnectBlock`: 
```cpp
validation.cpp:2136:35: warning: calling function 'CheckInputScripts' requires holding mutex 'cs_new' exclusively [-Wthread-safety-analysis]
            if (fScriptChecks && !CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], g_parallel_script_checks ? &vChecks : nullptr)) {
```

but there is no warning from the call site from `CheckInputsFromMempoolAndCache` or `PolicyScriptChecks`. 

then you can apply the annotation to the declaration instead of the definition, and see that all the call sites generate warnings.

File: src/validation.cpp [hunk 1]
```diff
@@ -198,7 +198,7 @@ CBlockIndex* BlockManager::FindForkInGlobalIndex(const CChain& chain, const CBlo
 
 std::unique_ptr<CBlockTreeDB> pblocktree;
 
-bool CheckInputScripts(const CTransaction& tx, TxValidationState &state, const CCoinsViewCache &inputs, unsigned int flags, bool cacheSigStore, bool cacheFullScriptStore, PrecomputedTransactionData& txdata, std::vector<CScriptCheck> *pvChecks = nullptr);
+bool CheckInputScripts(const CTransaction& tx, TxValidationState& state, const CCoinsViewCache& inputs, unsigned int flags, bool cacheSigStore, bool cacheFullScriptStore, PrecomputedTransactionData& txdata, std::vector<CScriptCheck>* pvChecks = nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Comment (hebasto) [inline src/validation.cpp:201]: What about another declaration https://github.com/bitcoin/bitcoin/blob/36be9b821ae0882ba5ccba543f959636e413991d/src/test/txvalidationcache_tests.cpp#L16?

File: doc/developer-notes.md [hunk 2]
```diff
@@ -785,6 +785,11 @@ Threads and synchronization
   get compile-time warnings about potential race conditions in code. Combine annotations in function declarations with
   run-time asserts in function definitions:
 
+  - In functions that are declared seperately from where they are defined, the
```
Comment (hebasto) [inline doc/developer-notes.md:788]: https://cirrus-ci.com/task/6341367142023168?command=lint#L859:
```
doc/developer-notes.md:788: seperately ==> separately
```

**Comment by hebasto** [comment]
If this pr a case to apply #20986?

File: src/validation.cpp (hunk 1 above)
Comment (amitiuttarwar) [reply src/validation.cpp]: good call, added annotation there too.

File: doc/developer-notes.md (hunk 2 above)
Comment (amitiuttarwar) [reply doc/developer-notes.md]: fixed

**Comment by amitiuttarwar** [comment]
thanks for the review @hebasto, took all your suggestions.

> If this pr a case to apply #20986?

done

**Review by promag — COMMENTED** [review]
Code review ACK 25c57d640992255ed67964a44b17afbfd4bed0cf.

How about adding a validation_p.h with these "private" declarations?

**Comment by MarcoFalke** [comment]
ACK 25c57d640992255ed67964a44b17afbfd4bed0cf 🥘

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 25c57d640992255ed67964a44b17afbfd4bed0cf 🥘
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUghXQwAxK2rZIu0qYdOFsfc6ncH58m6umvxB6lxhN7rpYnpz1AyiCe4+RbYe2Nx
ty820uHzkkQSb7eu40NeO7OkqcjlXHeUDSzXzzReSbOi90OLJa5pX4vMiur4chlP
AiGAQLuHhDtDOwOySpVBR7NxHkcaCUjD5CrGkbBieLaTJlmYcwqpI4hycCAR9A4h
Yg2hfcpwosTyUA7NNrk8F50UK7Ez97slqvDd7CLcI70jvEGpMiGO/YW1w2HqgJkP
s6x7l0g3Frd9q7VA/EucZoMZYbDnoWEGyP88UaT2G727WxuYhRrUDVVP7srND7Z1
JSNSzm/4v9eJtfaD0D/9xcgN0YxF6DRJqyJK6fPJAMdoimvpFIqyg2NowRaRcvHz
zTTTcygMRV6l/HMDNdNJ59VwWlvxE3MHu4yzSntRtgNaeW/a3bTnm4XZAcF5woQZ
2QhiNmS8BmHj2URMXcrNJmUmnWPj4bHaZuOhqIhvr8UMPeBhzTAzyT6Pyn3xU7Hs
4pT3enGx
=pXvj
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `29b3e3a94f59f0a1d3ed2364dbf2aa60e4373ca5a3c8c71576d43351cf60215b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010829b3e3a94f59f0a1d3ed2364dbf2aa60e4373ca5a3c8c71576d43351cf60215bf01075a731ef37771a4c1c24f791689e7ddf08fff01058e0841669dbe8afb6a8abb70d5c7f3a08f10460336f5ef0083c8a635bdbd19f5d0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010801210dd47d6d1420418f3e4863eca1708f02012f0da60823460193be3c59a1dd743ff89e4e14ae930e5a872be4643d1a289a908f02065537c9b93bf2528b7db0a3145e97b17b48f7179870e22b0957b2aaef0d30ff908f10460336f5ef0085dbb11e2438524020083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d7a7e9f2d0b2faeb7f6c54b6602e75ed08f10460336f5ef0084f640f66c34229af0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010dcb36c188bf5e026dcb3d5bbf9eb4b3a08f1046033618bf008e4e64f85b18f85c10083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Comment by amitiuttarwar** [comment]
@promag where are other examples of forward declaring functions from the tests? I see examples for other files eg. `denialofservice_tests` have declarations for orphan handling stuff from `net_processing`. But I'm not seeing other declarations for internal `validation.cpp` functions. I'm definitely open to the pattern of test-only header files, but in this case, seems like it might be a bit of overkill to have an additional header file for just one function? Unless I'm missing something?

**Pull request merged**